Add tests for omnisci client operations

dchigarev opened the pull request:
Add some tests from #1492.
 

1. `test_drop_table`
2. `test_create_table`
3. `test_truncate_table`
4. `test_create_view`
5. `test_drop_view`
6. `test_create_user`
7. `test_alter_user`
8. `test_create_database`
9. `test_drop_database`

dchigarev commented:
@xmnlab  PR is ready for review

datapythonista commented on ibis/omniscidb/tests/conftest.py:
```diff
@@ -200,3 +212,77 @@ def temp_database(con, test_data_db: str) -> str:
+        Random view name for a temporary usage.
+    """
+    name = _random_identifier('view')
+    try:
```
In `temp_table` you're dropping the table before returning the name, but not here for the view. Any reason for the inconsitency?

dchigarev replied:
fixed

datapythonista commented on ibis/omniscidb/tests/conftest.py:
```diff
@@ -200,3 +212,77 @@ def temp_database(con, test_data_db: str) -> str:
+
+@pytest.fixture(scope='function')
+def test_view(con, test_table) -> ibis.expr.types.TableExpr:
+    """Create a temporary view.
```
There is a lot of repetitive code in these functions, `temp_table`, `temp_view`, `test_view`...

Do you think would make sense to create a generic function to avoid the repetition? Something like:
```python
def test_helper(con, kind='table', create=False):
    ...

@pytest.fixture
def temp_view(con):
    return test_helper(con, kind='view', create=False)
```

Or probably something even better can be done, but personally I think the fixtures part starts to be too long and repetitive.

dchigarev replied:
done

pep8speaks commented:
Hello @dchigarev! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`ibis/omniscidb/tests/test_client.py`](https://github.com/ibis-project/ibis/blob/f27d14a485637a73398645c8396223c1593c2047/ibis/omniscidb/tests/test_client.py):

> [Line 139:80](https://github.com/ibis-project/ibis/blob/f27d14a485637a73398645c8396223c1593c2047/ibis/omniscidb/tests/test_client.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)

datapythonista commented on ibis/omniscidb/tests/conftest.py:
```diff
@@ -36,26 +36,70 @@ def con():
+    -------
+    ibis.expr.types.TableExpr
+    """
+    with temp_helper(con, 'table', True, {'schema': test_schema}) as name:
```
For things that are not obvious, I think it's better to use the parameter name. Reading this, it's difficult to guess what's `True` without having to go to the function definintion. I think being explicit helps readability:
```suggestion
    with temp_helper(con, kind='table', create=True, create_kwargs={'schema': test_schema}) as name:
```

dchigarev replied:
done

dchigarev commented:
@datapythonista 
> Thanks for the changes, I think this looks much better than repeating everything.
> 
> I'm wondering it it's worth having functions for `test_table`, `temp_view`... I would probably use the context manager directly, something like:
> 
> ```python
> import contextlib
> 
> @contextlib.contextmanager
> def create_db_obj(kind):
>     print(f'CREATE obj_name')
>     yield 'obj_name'
>     print(f'DROP obj_name')
> 
> def test_whatever():
>     with create_db_obj('view') as view_name:
>         print(f'SELECT * FROM {view_name}')
> 
> test_whatever()
> ```
> 
> What do you think?

Well, for now we have `test_table`, `temp_view` etc, as a fixtures, so we pass them in a test functions as a params. With the new approach, we gotta change all test functions to manually create temp entities inside them for their purposes. Should we do that?

datapythonista commented:
> Well, for now we have `test_table`, `temp_view` etc, as a fixtures, so we pass them in a test functions as a params. With the new approach, we gotta change all test functions to manually create temp entities inside them for their purposes. Should we do that?

It's just an idea, using fixtures seems also reasonable.

CI seems to be red, if you want to have a look...

dchigarev commented:
hm, it seems it is CI issue, because it crashes with `no space left on device` error, I'll try to merge that branch with master and see if it helps

dchigarev commented:
@datapythonista well, now CI is green

datapythonista commented:
We had to disable omnisci tests in the CI to fix it. I'm trying to restore them in #2205, but I'm getting some weird errors from pytest. We need that PR merged before we can merge this. If this is breaking omnisci tests, the CI won't tell us.

dchigarev commented:
Now CI is green, including omnisci tests

datapythonista commented:
@jreback, I see I can't still merge (not that I would merge this without your review).

I see that I'm not part of the `ibis-project` organization. Not sure if it can be related?

dchigarev commented:
@datapythonista @jreback just friendly reminder about this PR)

jreback commented on ibis/omniscidb/tests/conftest.py:
```diff
@@ -36,26 +36,72 @@ def con():
     )
 
 
+class temp_helper:
```
i get what you are trying to do. but i much prefer very explict operations in tests. I am ok with having a base class for some of this, but each operation should be a named function.

also classes need to be CamelCase generally.

dchigarev replied:
I'm not sure what do you mean under named function for each operation.

also changed class name to CamelCase style

jreback commented on ibis/omniscidb/tests/conftest.py:
```diff
@@ -193,10 +237,56 @@ def temp_database(con, test_data_db: str) -> str:
+        yield ret_value
+
+
+@pytest.fixture
```
so I actually don't like this temp_* at all. This is very hard to read, much better to just use: 'table', 'view', 'user' etc. These are fixtures after all.

dchigarev replied:
done it in the last commit

jreback commented on ibis/omniscidb/tests/conftest.py:
```diff
@@ -193,10 +237,56 @@ def temp_database(con, test_data_db: str) -> str:
+        yield ret_value
+
+
+@pytest.fixture(scope='function')
```
these default to function scope already. However, is there a reason we have function scope? aren't these stateful resources? This is part of the reason you have to jump thru hoops to any kind of testing.

I would prefer if you are refactoring anyhow to treat these as proper resources see: https://docs.pytest.org/en/stable/fixture.html

dchigarev replied:
these fixtures creates a temporary object for testing (tables, views, databases, ...), and since we're dropping these objects in some tests (for example in `drop_*`) we can't set scope that wider than `function`

datapythonista commented:
@dchigarev can you merge master, fix conflicts and have a look at the comments from the last review please?

dchigarev commented:
@datapythonista @jreback merged master and moved some tests from only OmniSci tests to general tests, also renamed some fixtures entities to make their names more explicit

jreback commented on ibis/omniscidb/tests/test_client.py:
```diff
@@ -287,3 +287,97 @@ def test_cpu_execution_type(
+
+    if isinstance(table_src, ibis.Schema):
+        con.create_table(table_name, schema=table_src)
+    else:
```
why do you have this path in here? this hits the 2 xfailed ones right? i would simply make that a different test as this is *very* confusing.

jreback commented on ibis/omniscidb/tests/test_client.py:
```diff
@@ -287,3 +287,97 @@ def test_cpu_execution_type(
+            user=name,
+            password=password,
+        )
+    except Exception:
```
huh? what exactly is this testing? why do you have a try/except around it in a test? if you really need to test something like this, pls use a mock

jreback commented on ibis/omniscidb/tests/test_client.py:
```diff
@@ -287,3 +287,97 @@ def test_cpu_execution_type(
+            port=con.port,
+            user=name,
+            password=password,
+        )
```
same catching exceptions inside tests just hides things

jreback commented on ibis/tests/all/conftest.py:
```diff
@@ -377,46 +373,71 @@ def get_common_spark_testing_client(data_directory, connect):
     Parameters
     ----------
-    con : ibis.client.Client
+    con : ibis.omniscidb.OmniSciDBClient
```
revert this

jreback commented on ibis/tests/all/test_client.py:
```diff
@@ -14,6 +15,20 @@
 )
 
 
+def _drop(con, name, method_name, drop_kwargs, create_kwargs):
```
instead of this, can you simply create a fixture which does this? it is not at all obvious *where* this is used (yes i can see it), but it is magical in this function and is mutating db state.

jreback commented on ibis/tests/all/test_client.py:
```diff
@@ -160,40 +175,103 @@ def test_nullable_input_output(con, backend, temp_table):
+    try:
+        con.create_view(view, alltypes[expr])
+    except com.IbisTypeError:
+        assert not set(expr).issubset(df_alltypes.columns)
```
instead of try/excepting, can you differeniate the list into those which work and those which are supposed to fail.

jreback commented on ibis/tests/all/test_client.py:
```diff
@@ -160,40 +175,103 @@ def test_nullable_input_output(con, backend, temp_table):
+    # if an exception occurred with force=False
+    try:
+        con.drop_view(new_view, force=force)
+    except Exception:
```
pls pls don't try/except inside tests, use a fixture instead

jreback commented on ibis/tests/util.py:
```diff
@@ -15,3 +15,39 @@ def assert_equal(left, right):
+
+
+class TempHelper:
+    def __init__(
```
can you add an explanation here

datapythonista commented:
@dchigarev do you have time to address the comments from the review?

datapythonista commented:
Closing as stale, please let me know if you want to continue working on this @dchigarev, and I'll reopen.
